### PR TITLE
feat(components, app, labware-library): migrate to react-router-dom v6.24.1

### PR DESCRIPTION
**Pull request by jerader**
closes AUTH-564

# Overview

This PR updates `react-router-dom` to v6.24.1 in the stack to match PD's `react-router-dom` version

Big changes between v5 and v6:

1. `useHistory` is deprecated and replaced with `useNavigate`. Instead of `history.push('path')` and `history.goBack()` it is `navigate('path')` and `navigate(-1)`
2.  `Route` no longer has children, instead the component goes into the `element` prop. 
3. `Switch` is deprecated and is now `Routes`
4. `Route` must be wrapped in `Routes` (should mainly have to worry about this in unit testing since the app components are wrapped in `Routes` already from the top level)
5. `<Redirect/>` is deprecated and is now `<Navigate/>` (if it is wrapped in `Routes` then you may use `Route` and `Navigate` goes into the element prop)

# Test Plan

Smoke test labware-library, protocol designer but most importantly, smoke test the app and ODD.

# Changelog

- update `react-router-dom` in Labware-library and app
- deprecate it in `components`
- fix all affected components and tests

# Review requests

see test plan

# Risk assessment

high since we have a 8.0.0 release and this touches a bunch of code. we need to smoke test well before merging

### CODE REVIEW

File: app/src/index.tsx [hunk 1]
```diff
@@ -20,6 +17,7 @@ import '../src/atoms/SoftwareKeyboard/NumericalKeyboard/index.css'
 
 // component tree
 import { App } from './App'
+import { BrowserRouter } from 'react-router-dom'
```
Comment (koji) [inline app/src/index.tsx]: this line can be after `react-redux`

**Comment by koji** [comment]
the migration guide would be useful to review this pr
https://github.com/remix-run/react-router/discussions/8753

File: app/src/organisms/Devices/RobotSettings/AdvancedTab/AdvancedTabSlideouts/DeviceResetModal.tsx [hunk 2]
```diff
@@ -79,7 +79,7 @@ export function DeviceResetModal({
         }
       }
       dispatchRequest(resetConfig(robotName, resetOptions))
-      history.push(`/devices/`)
+      navigate(`/devices/`)
```
Comment (koji) [inline app/src/organisms/Devices/RobotSettings/AdvancedTab/AdvancedTabSlideouts/DeviceResetModal.tsx]:
```suggestion
      navigate('/devices/')
```

File: app/src/organisms/InstrumentMountItem/AttachedInstrumentMountItem.tsx [hunk 3]
```diff
@@ -101,7 +100,7 @@ export function AttachedInstrumentMountItem(
                 setShowChoosePipetteModal(false)
               },
               onComplete: () => {
-                history.push(
+                navigate(
                   attachedInstrument == null
                     ? `/instruments`
```
Comment (koji) [inline app/src/organisms/InstrumentMountItem/AttachedInstrumentMountItem.tsx:105]:
```suggestion
                    ? '/instruments'
```

File: app/src/organisms/InstrumentMountItem/AttachedInstrumentMountItem.tsx [hunk 4]
```diff
@@ -53,7 +52,7 @@ export function AttachedInstrumentMountItem(
         flowType: GRIPPER_FLOW_TYPES.ATTACH,
         attachedGripper: attachedInstrument,
         onComplete: () => {
-          history.push(
+          navigate(
             attachedInstrument == null ? `/instruments` : `/instrument/${mount}`
```
Comment (koji) [inline app/src/organisms/InstrumentMountItem/AttachedInstrumentMountItem.tsx]:
```suggestion
            attachedInstrument == null ? '/instruments' : `/instrument/${mount}`
```

File: app/src/organisms/Navigation/__tests__/NavigationMenu.test.tsx [hunk 5]
```diff
@@ -9,19 +9,19 @@ import { useLights } from '../../Devices/hooks'
 import { RestartRobotConfirmationModal } from '../RestartRobotConfirmationModal'
 import { NavigationMenu } from '../NavigationMenu'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/Navigation/__tests__/NavigationMenu.test.tsx]: https://reactrouter.com/en/main/hooks/use-navigate#usenavigate


```suggestion
import type { NavigateFunction } from 'react-router-dom'
```

File: app/src/organisms/ChangePipette/__tests__/ChangePipette.test.tsx [hunk 6]
```diff
@@ -21,18 +21,18 @@ import { ExitModal } from '../ExitModal'
 import { ConfirmPipette } from '../ConfirmPipette'
 import { ChangePipette } from '..'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/ChangePipette/__tests__/ChangePipette.test.tsx]: same 
recommend you use NavigationFunction type

File: app/src/organisms/DeviceDetailsDeckConfiguration/__tests__/DeckConfigurationDiscardChangesModal.test.tsx [hunk 7]
```diff
@@ -5,16 +5,16 @@ import { describe, it, beforeEach, vi, expect } from 'vitest'
 import { renderWithProviders } from '../../../__testing-utils__'
 import { i18n } from '../../../i18n'
 import { DeckConfigurationDiscardChangesModal } from '../DeckConfigurationDiscardChangesModal'
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/DeviceDetailsDeckConfiguration/__tests__/DeckConfigurationDiscardChangesModal.test.tsx]: same

File: app/src/organisms/NetworkSettings/__tests__/AlternativeSecurityTypeModal.test.tsx [hunk 8]
```diff
@@ -6,15 +6,15 @@ import { renderWithProviders } from '../../../__testing-utils__'
 import { i18n } from '../../../i18n'
 import { AlternativeSecurityTypeModal } from '../AlternativeSecurityTypeModal'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/NetworkSettings/__tests__/AlternativeSecurityTypeModal.test.tsx]: same

File: app/src/organisms/NetworkSettings/__tests__/DisplayWifiList.test.tsx [hunk 9]
```diff
@@ -8,9 +8,9 @@ import * as Fixtures from '../../../redux/networking/__fixtures__'
 import { DisplaySearchNetwork } from '../DisplaySearchNetwork'
 import { DisplayWifiList } from '../DisplayWifiList'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/NetworkSettings/__tests__/DisplayWifiList.test.tsx]: same

File: app/src/organisms/NetworkSettings/__tests__/SelectAuthenticationType.test.tsx [hunk 10]
```diff
@@ -11,9 +11,9 @@ import { AlternativeSecurityTypeModal } from '../AlternativeSecurityTypeModal'
 import { SelectAuthenticationType } from '../SelectAuthenticationType'
 import { SetWifiCred } from '../SetWifiCred'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/NetworkSettings/__tests__/SelectAuthenticationType.test.tsx]: same

File: app/src/organisms/NetworkSettings/__tests__/WifiConnectionDetails.test.tsx [hunk 11]
```diff
@@ -11,19 +11,19 @@ import * as Fixtures from '../../../redux/networking/__fixtures__'
 import { NetworkDetailsModal } from '../../RobotSettingsDashboard/NetworkSettings/NetworkDetailsModal'
 import { WifiConnectionDetails } from '../WifiConnectionDetails'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/NetworkSettings/__tests__/WifiConnectionDetails.test.tsx]: same

File: app/src/organisms/OnDeviceDisplay/NameRobot/__tests__/ConfirmRobotName.test.tsx [hunk 12]
```diff
@@ -7,15 +7,15 @@ import { renderWithProviders } from '../../../../__testing-utils__'
 import { i18n } from '../../../../i18n'
 import { ConfirmRobotName } from '../ConfirmRobotName'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/OnDeviceDisplay/NameRobot/__tests__/ConfirmRobotName.test.tsx]: same

File: app/src/organisms/OnDeviceDisplay/RobotDashboard/__tests__/RecentRunProtocolCard.test.tsx [hunk 13]
```diff
@@ -33,16 +33,16 @@ import {
   INIT_STATUS,
 } from '../../../../resources/health/hooks'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/OnDeviceDisplay/RobotDashboard/__tests__/RecentRunProtocolCard.test.tsx]: same

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/__tests__/ConfirmCancelRunModal.test.tsx [hunk 14]
```diff
@@ -20,7 +20,7 @@ import { mockConnectedRobot } from '../../../../redux/discovery/__fixtures__'
 import { ConfirmCancelRunModal } from '../ConfirmCancelRunModal'
 import { CancelingRunModal } from '../CancelingRunModal'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/OnDeviceDisplay/RunningProtocol/__tests__/ConfirmCancelRunModal.test.tsx]: same

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/__tests__/RunFailedModal.test.tsx [hunk 15]
```diff
@@ -9,13 +9,13 @@ import { renderWithProviders } from '../../../../__testing-utils__'
 import { i18n } from '../../../../i18n'
 import { RunFailedModal } from '../RunFailedModal'
 
-import type { useHistory } from 'react-router-dom'
+import type { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/OnDeviceDisplay/RunningProtocol/__tests__/RunFailedModal.test.tsx]: same

File: app/package.json [hunk 16]
```diff
@@ -53,7 +53,7 @@
     "react-intersection-observer": "^8.33.1",
     "react-markdown": "9.0.1",
     "react-redux": "8.1.2",
-    "react-router-dom": "5.3.4",
+    "react-router-dom": "6.24.1",
```
Comment (koji) [inline app/package.json:56]: not introducing v6.25.0/v6.25.1 because too new?

File: app/src/molecules/CardButton/__tests__/CardButton.test.tsx [hunk 17]
```diff
@@ -9,13 +9,13 @@ import { i18n } from '../../../i18n'
 import { CardButton } from '..'
 import type * as ReactRouterDom from 'react-router-dom'
```
Comment (koji) [inline app/src/molecules/CardButton/__tests__/CardButton.test.tsx]: maybe use `NavigateFunction` type?

File: app/src/organisms/ModuleCard/__tests__/ModuleCard.test.tsx [hunk 18]
```diff
@@ -37,6 +37,7 @@ import { FirmwareUpdateFailedModal } from '../FirmwareUpdateFailedModal'
 import { ErrorInfo } from '../ErrorInfo'
 import { ModuleCard } from '..'
 
+import type * as ReactRouterDom from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/ModuleCard/__tests__/ModuleCard.test.tsx]: can we use `NavigateFuntion` type here?

File: app/src/organisms/ProtocolSetupModulesAndDeck/index.tsx [hunk 19]
```diff
@@ -36,7 +36,7 @@ import type { CutoutId, CutoutFixtureId } from '@opentrons/shared-data'
 import type { SetupScreens } from '../../pages/ProtocolSetup'
 import { useRunStatus } from '../RunTimeControl/hooks'
 import { RUN_STATUS_STOPPED } from '@opentrons/api-client'
-import { useHistory } from 'react-router-dom'
+import { useNavigate } from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/ProtocolSetupModulesAndDeck/index.tsx]: this line can be moved to line 4
also could you move the above 2 lines too?

File: app/src/organisms/OnDeviceDisplay/RunningProtocol/ConfirmCancelRunModal.tsx [hunk 20]
```diff
@@ -75,9 +75,9 @@ export function ConfirmCancelRunModal({
       dismissCurrentRun(runId)
       if (!isActiveRun) {
         if (protocolId != null) {
-          history.push(`/protocols/${protocolId}`)
+          navigate(`/protocols/${protocolId}`)
         } else {
-          history.push(`/protocols`)
+          navigate(`/protocols`)
```
Comment (koji) [inline app/src/organisms/OnDeviceDisplay/RunningProtocol/ConfirmCancelRunModal.tsx]:
```suggestion
          navigate('/protocols')
```

File: app/src/organisms/QuickTransferFlow/__tests__/SummaryAndSettings.test.tsx [hunk 21]
```diff
@@ -14,13 +14,13 @@ import { NameQuickTransfer } from '../NameQuickTransfer'
 import { Overview } from '../Overview'
 import type * as ReactRouterDom from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/QuickTransferFlow/__tests__/SummaryAndSettings.test.tsx]: can we use `NavigateFuntion`?

File: app/src/organisms/RobotSettingsDashboard/__tests__/RobotSystemVersionModal.test.tsx [hunk 22]
```diff
@@ -9,13 +9,13 @@ import { RobotSystemVersionModal } from '../RobotSystemVersionModal'
 import type * as Dom from 'react-router-dom'
```
Comment (koji) [inline app/src/organisms/RobotSettingsDashboard/__tests__/RobotSystemVersionModal.test.tsx:9]: can we use NavigateFuntion?

File: app/package.json (hunk 16 above)
Comment (jerader) [reply app/package.json:56]: oh wait i think v6.25.0 came out just yesterday, that is why i did not use it lol. Shouldn't we stick with 6.24.1 to match PD though? hmm. @shlokamin thoughts?

File: app/src/pages/InstrumentsDashboard/__tests__/InstrumentsDashboard.test.tsx [hunk 23]
```diff
@@ -12,6 +12,7 @@ import { InstrumentsDashboard } from '..'
 import { formatTimeWithUtcLabel } from '../../../resources/runs'
 import { InstrumentDetail } from '../../../pages/InstrumentDetail'
 import type * as ReactApiClient from '@opentrons/react-api-client'
+import type * as ReactRouterDom from 'react-router-dom'
```
Comment (koji) [inline app/src/pages/InstrumentsDashboard/__tests__/InstrumentsDashboard.test.tsx]:
```suggestion

```

File: app/package.json (hunk 16 above)
Comment (koji) [reply app/package.json:56]: 6.25.0 2days ago and 6.25.1 yesterday.
https://github.com/remix-run/react-router/releases

File: app/src/App/OnDeviceDisplayApp.tsx [hunk 24]
```diff
@@ -252,23 +252,31 @@ export function OnDeviceDisplayAppRoutes(): JSX.Element {
...
+              {getPathComponent(path)}
+            </Box>
+          }
+        />
       ))}
-      {targetPath != null && <Redirect exact from="/" to={targetPath} />}
-    </Switch>
+      {targetPath != null && (
+        <Route path="/" element={<Navigate to={targetPath} />} />
+      )}
```
Comment (koji) [inline app/src/App/OnDeviceDisplayApp.tsx]: https://stackoverflow.com/questions/69868956/how-can-i-redirect-in-react-router-v6

```suggestion
      <Route path="*" element={<Navigate to={targetPath} replace />} />
```

**Comment by koji** [comment]
For ODD, we would need to check the followings
unboxing flow [without software update]
welcome -> network setup -> software update -> e-stop screen -> name robot -> finish setup -> robot dashboard with modal

unboxing flow [with software update]
welcome -> network setup -> software update -> reboot the robot -> e-stop screen -> name robot -> finish setup -> robot dashboard with modal


Navigation item after unboxing flow 
- protocols
- quick transfer
- instruments
- settings

Robot settings Dashboard
- network setup flow
- rename flow

File: app/src/App/OnDeviceDisplayApp.tsx (hunk 24 above)
Comment (jerader) [reply app/src/App/OnDeviceDisplayApp.tsx]: thank you so much koji!

File: app/package.json (hunk 16 above)
Comment (shlokamin) [reply app/package.json:56]: doesn't matter to me as long as all of our frontend projects share the same version, so we can either update PD or keep these at 6.24.1

**Comment by shlokamin** [comment]
making my way through this PR but chiming in to say that i can devote time early next week to test this on a bot!

**Review by koji — APPROVED** [review]
Tested this branch on ODD (tested the UnboxingFlow 2,3 times a couple of times) and didn't see any issues.
I think once Shlok tested this branch and if he doesn't see any issue, this branch will be good to go!

Thank you for updating `react-router`!

**Pull request merged**